File: packages/cli/src/sources/npm.ts
import type {
  DocFile,
  DocSource,
  FetchResult,
  NpmSourceOptions,
  SourceConfig,
} from './index.js'
import { execSync } from 'node:child_process'
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import process from 'node:process'
import { consola } from 'consola'
import { satisfies, validRange } from 'semver'
import {
  acquireEntryLock,
  npmStorePath,
  resolveAskHome,
  stampEntry,
  writeEntryAtomic,
} from '../store/index.js'

/**
 * npm source — fetches docs from a published tarball OR, when the package is
 * already installed in the current project, directly from
 * `node_modules/<pkg>/<docsPath>`.
 *
 * Local-first behavior was added in npm-tarball-docs-20260408 to (1) avoid
 * unnecessary network IO when the docs are already present on disk and
 * (2) keep AI agents working offline / inside CI sandboxes. The local path
 * is taken whenever the installed `package.json` version satisfies the
 * requested version (or the request is `latest`) AND the configured
 * `<docsPath>` exists. Otherwise the existing tarball download path runs.
 */
export class NpmSource implements DocSource {
  async fetch(options: SourceConfig): Promise<FetchResult> {
    const opts = options as NpmSourceOptions
    const pkg = opts.package ?? opts.name

    const local = this.tryLocalRead({
      projectDir: process.cwd(),
      pkg,
      requestedVersion: opts.version,
      docsPath: opts.docsPath,
    })
    if (local) {
      consola.info(`  Using local node_modules for ${pkg}@${local.resolvedVersion}`)
      // Also materialize into the global store for cross-project reuse
      await this.writeToStore(pkg, local.resolvedVersion, local.files)
      const askHome = resolveAskHome()
      local.storePath = npmStorePath(askHome, pkg, local.resolvedVersion)
      return local
    }

    return this.fetchFromTarball(opts, pkg)
  }

  /**
   * Read docs directly from `node_modules/<pkg>` when the installed version
   * satisfies the request and the configured `docsPath` exists. Returns
   * `null` to indicate the local path is not viable — the caller falls back
   * to the tarball download.
   *
   * Public-ish (not exported) for unit testing via `(new NpmSource()).tryLocalRead`.
   */
  tryLocalRead(args: {
    projectDir: string
    pkg: string
    requestedVersion: string
    docsPath?: string
  }): FetchResult | null {
    const { projectDir, pkg, requestedVersion, docsPath } = args

    if (!docsPath) {
      // No explicit docsPath: we don't try to auto-detect against
      // node_modules. The local path exists for the curated case.
      return null
    }

    // Resolve the package's installed root. Scoped packages (`@scope/pkg`)
    // live at `node_modules/@scope/pkg/`.
    const pkgDir = path.join(projectDir, 'node_modules', pkg)
    const pkgJsonPath = path.join(pkgDir, 'package.json')
    if (!fs.existsSync(pkgJsonPath)) {
      return null
    }

    let installedVersion: string
    try {
      const meta = JSON.parse(fs.readFileSync(pkgJsonPath, 'utf-8')) as { version?: string }
      if (!meta.version) {
        return null
      }
      installedVersion = meta.version
    }
    catch {
      return null
    }

    if (!this.versionMatches(requestedVersion, installedVersion)) {
      return null
    }

    // Defense-in-depth: even though `docsPath` originates from a trusted
    // registry strategy, treat it as untrusted and reject anything that
    // resolves outside the installed package directory. A malformed entry
    // like `docsPath: '../../../etc/passwd'` would otherwise read arbitrary
    // files on local-first reads.
    //
    // Two-stage check:
    //   1. String-level: reject obvious traversal (`..`, absolute path)
    //      before touching the filesystem.
    //   2. Realpath: after confirming `docsDir` exists, resolve symlinks on
    //      both `pkgDir` and `docsDir` and re-check containment. A symlink
    //      inside the package dir that points outside (e.g.
    //      `node_modules/<pkg>/dist/docs -> /etc`) bypasses the string
    //      check but is caught here.
    const docsDir = path.resolve(pkgDir, docsPath)
    const relativeDocsDir = path.relative(pkgDir, docsDir)
    if (relativeDocsDir.startsWith('..') || path.isAbsolute(relativeDocsDir)) {
      return null
    }
    if (!fs.existsSync(docsDir)) {
      return null
    }
    let realPkgDir: string
    let realDocsDir: string
    try {
      realPkgDir = fs.realpathSync(pkgDir)
      realDocsDir = fs.realpathSync(docsDir)
    }
    catch {
      return null
    }
    const realRelative = path.relative(realPkgDir, realDocsDir)
    if (realRelative.startsWith('..') || path.isAbsolute(realRelative)) {
      return null
    }

    let files: DocFile[]
    if (fs.statSync(docsDir).isFile()) {
      const content = fs.readFileSync(docsDir, 'utf-8')
      files = [{ path: path.basename(docsDir), content }]
    }
    else {
      files = this.collectMarkdownFiles(docsDir, docsDir)
    }

    if (files.length === 0) {
      // The directory exists but holds no readable docs — treat as a miss
      // so we don't poison the lockfile with an empty entry.
      return null
    }

    return {
      files,
      resolvedVersion: installedVersion,
      meta: { installPath: pkgDir },
    }
  }

  /**
   * Match policy for the local-first read:
   *   - `latest`        → any installed version is acceptable
   *   - exact version   → must equal installed
   *   - semver range    → installed must satisfy
   *   - non-semver tag  → must equal installed (treat as opaque)
   *
   * `requestedVersion` is whatever the caller passed in (`SourceConfig.version`).
   * It originates from the registry resolution + manifest gate, so by the
   * time we get here it is usually an exact version. The semver-range branch
   * is defensive cover for callers that haven't run the gate.
   */
  private versionMatches(requested: string, installed: string): boolean {
    if (requested === 'latest') {
      return true
    }
    if (validRange(requested)) {
      return satisfies(installed, requested)
    }
    return requested === installed
  }

  private async fetchFromTarball(opts: NpmSourceOptions, pkg: string): Promise<FetchResult> {
    const spec = `${pkg}@${opts.version}`

    // Get package info to resolve exact version and tarball URL
    const resolvedVersion = execSync(`npm view ${spec} version`, {
      encoding: 'utf-8',
    }).trim()

    const tarballUrl = execSync(`npm view ${spec} dist.tarball`, {
      encoding: 'utf-8',
    }).trim()

    const integrity = (() => {
      try {
        return execSync(`npm view ${spec} dist.integrity`, {
          encoding: 'utf-8',
        }).trim() || undefined
      }
      catch {
        return undefined
      }
    })()

    // Download and extract tarball
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'ask-npm-'))

    try {
      consola.info(`  Downloading ${pkg}@${resolvedVersion}...`)
      execSync(`curl -sL "${tarballUrl}" | tar xz -C "${tmpDir}"`, {
        encoding: 'utf-8',
      })

      // Find docs in the extracted package
      const packageDir = path.join(tmpDir, 'package')
      const docsPath = opts.docsPath ?? this.detectDocsPath(packageDir)

      if (!docsPath) {
        throw new Error(
          `No docs found in ${spec}. Specify --docs-path to point to the docs directory within the package.`,
        )
      }

      const docsDir = path.join(packageDir, docsPath)
      if (!fs.existsSync(docsDir)) {
        throw new Error(
          `Docs path "${docsPath}" not found in ${spec}. Available paths:\n${this.listDirs(packageDir)}`,
        )
      }

      let files: DocFile[]
      if (fs.statSync(docsDir).isFile()) {
        const content = fs.readFileSync(docsDir, 'utf-8')
        files = [{ path: path.basename(docsDir), content }]
      }
      else {
        files = this.collectMarkdownFiles(docsDir, docsDir)
      }
      await this.writeToStore(pkg, resolvedVersion, files)
      const askHome = resolveAskHome()

      return {
        files,
        resolvedVersion,
        storePath: npmStorePath(askHome, pkg, resolvedVersion),
        meta: { tarball: tarballUrl, integrity },
      }
    }
    finally {
      fs.rmSync(tmpDir, { recursive: true, force: true })
    }
  }

  private detectDocsPath(packageDir: string): string | null {
    const candidates = [
      'docs',
      'doc',
      'dist/docs',
      'documentation',
      'guide',
      'guides',
    ]
    for (const candidate of candidates) {
      const fullPath = path.join(packageDir, candidate)
      if (fs.existsSync(fullPath) && fs.statSync(fullPath).isDirectory()) {
        return candidate
      }
    }
    return null
  }

  private listDirs(dir: string): string {
    return fs
      .readdirSync(dir, { withFileTypes: true })
      .filter(d => d.isDirectory())
      .map(d => `  - ${d.name}/`)
      .join('\n')
  }

  private collectMarkdownFiles(
    baseDir: string,
    currentDir: string,
  ): DocFile[] {
    const files: DocFile[] = []
    const entries = fs.readdirSync(currentDir, { withFileTypes: true })

    for (const entry of entries) {
      const fullPath = path.join(currentDir, entry.name)
      if (entry.isDirectory()) {
        files.push(...this.collectMarkdownFiles(baseDir, fullPath))
      }
      else if (this.isDocFile(entry.name)) {
        const relativePath = path.relative(baseDir, fullPath)
        const content = fs.readFileSync(fullPath, 'utf-8')
        files.push({ path: relativePath, content })
      }
    }

    return files
  }

  private isDocFile(filename: string): boolean {
    const ext = path.extname(filename).toLowerCase()
    return ['.md', '.mdx', '.txt', '.rst'].includes(ext)
  }

  /**
   * Write fetched docs into the global store at
   * `<ASK_HOME>/npm/<pkg>@<version>/`. Skips if the entry already exists.
   */
  private async writeToStore(
    pkg: string,
    version: string,
    files: DocFile[],
  ): Promise<void> {
    const askHome = resolveAskHome()
    const storeDir = npmStorePath(askHome, pkg, version)

    if (fs.existsSync(storeDir)) {
      return // already stored
    }

    const lock = await acquireEntryLock(storeDir)
    if (!lock) {
      return // another process completed the write
    }

    try {
      writeEntryAtomic(storeDir, files)
      stampEntry(storeDir)
    }
    finally {
      lock.release()
    }
  }
}

File: crates/ask/src/sources/npm.rs
//! npm source — local-first read from `node_modules/<pkg>/<docsPath>`, falling
//! back to a published-tarball download. Rust port of `sources/npm.ts`.

use std::io::Cursor;
use std::path::Path;
use std::process::Command;

use anyhow::{bail, Result};
use flate2::read::GzDecoder;

use super::{
    collect_doc_files, detect_docs_path, to_entry_files, DocFile, FetchMeta, FetchResult,
    NPM_DOC_CANDIDATES,
};
use crate::http::HttpClient;
use crate::store::{
    acquire_entry_lock, lexical_clean, npm_store_path, stamp_entry, write_entry_atomic,
};

/// Options for an npm fetch (port of `NpmSourceOptions`).
#[derive(Debug, Clone, Default)]
pub struct NpmOptions {
    pub name: String,
    pub version: String,
    pub package: Option<String>,
    pub docs_path: Option<String>,
}

/// Fetch docs for an npm package: try the installed `node_modules` copy first,
/// then a tarball download.
pub fn fetch(
    client: &dyn HttpClient,
    opts: &NpmOptions,
    project_dir: &Path,
    ask_home: &Path,
) -> Result<FetchResult> {
    let pkg = opts.package.as_deref().unwrap_or(&opts.name);

    if let Some(mut local) =
        try_local_read(project_dir, pkg, &opts.version, opts.docs_path.as_deref())
    {
        write_to_store(ask_home, pkg, &local.resolved_version, &local.files)?;
        local.store_path = Some(npm_store_path(ask_home, pkg, &local.resolved_version)?);
        return Ok(local);
    }

    fetch_from_tarball(client, opts, pkg, ask_home)
}

/// Match policy for the local-first read (node-semver semantics):
/// `latest` → any; a bare exact version → equality; a semver range → satisfies;
/// an opaque tag → equality.
fn version_matches(requested: &str, installed: &str) -> bool {
    if requested == "latest" {
        return true;
    }
    // A plain version string is an EXACT match in npm's range grammar (unlike the
    // Cargo `semver` crate, where `1.2.3` means `^1.2.3`), so compare directly.
    if semver::Version::parse(requested).is_ok() {
        return requested == installed;
    }
    if let (Ok(req), Ok(v)) = (
        semver::VersionReq::parse(requested),
        semver::Version::parse(installed),
    ) {
        return req.matches(&v);
    }
    requested == installed
}

/// Read docs directly from `node_modules/<pkg>` when the installed version
/// satisfies the request and the configured `docs_path` exists. `None` = not
/// viable (caller falls back to the tarball).
pub fn try_local_read(
    project_dir: &Path,
    pkg: &str,
    requested_version: &str,
    docs_path: Option<&str>,
) -> Option<FetchResult> {
    // No explicit docsPath: the local path exists only for the curated case.
    let docs_path = docs_path?;

    let pkg_dir = project_dir.join("node_modules").join(pkg);
    let pkg_json = pkg_dir.join("package.json");
    let meta_text = std::fs::read_to_string(&pkg_json).ok()?;
    let installed_version = serde_json::from_str::<serde_json::Value>(&meta_text)
        .ok()?
        .get("version")?
        .as_str()?
        .to_string();

    if !version_matches(requested_version, &installed_version) {
        return None;
    }

    // String-level traversal guard (before touching the fs).
    let docs_dir = pkg_dir.join(docs_path);
    if !lexical_clean(&docs_dir).starts_with(lexical_clean(&pkg_dir)) {
        return None;
    }
    if !docs_dir.exists() {
        return None;
    }
    // Realpath guard: a symlink inside the package pointing outside is caught here.
    let real_pkg = std::fs::canonicalize(&pkg_dir).ok()?;
    let real_docs = std::fs::canonicalize(&docs_dir).ok()?;
    if !real_docs.starts_with(&real_pkg) {
        return None;
    }

    let files = if docs_dir.is_file() {
        let content = std::fs::read_to_string(&docs_dir).ok()?;
        let name = docs_dir.file_name()?.to_string_lossy().into_owned();
        vec![DocFile {
            path: name,
            content,
        }]
    } else {
        collect_doc_files(&docs_dir, &docs_dir).ok()?
    };
    if files.is_empty() {
        return None;
    }

    Some(FetchResult {
        files,
        resolved_version: installed_version,
        store_path: None,
        store_subpath: None,
        from_store_cache: false,
        meta: FetchMeta {
            install_path: Some(pkg_dir),
            ..Default::default()
        },
    })
}

fn npm_view(spec: &str, field: &str) -> Result<String> {
    let output = Command::new("npm").args(["view", spec, field]).output()?;
    if !output.status.success() {
        bail!(
            "npm view {spec} {field} failed: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

fn fetch_from_tarball(
    client: &dyn HttpClient,
    opts: &NpmOptions,
    pkg: &str,
    ask_home: &Path,
) -> Result<FetchResult> {
    let spec = format!("{pkg}@{}", opts.version);
    let resolved_version = npm_view(&spec, "version")?;
    let tarball_url = npm_view(&spec, "dist.tarball")?;
    let integrity = npm_view(&spec, "dist.integrity")
        .ok()
        .filter(|s| !s.is_empty());

    let response = client.get_bytes(&tarball_url, &[])?;
    if !response.ok() {
        bail!("Failed to download {tarball_url}: HTTP {}", response.status);
    }

    let tmp = tempfile::Builder::new().prefix("ask-npm-").tempdir()?;
    let decoder = GzDecoder::new(Cursor::new(&response.body));
    tar::Archive::new(decoder).unpack(tmp.path())?;

    let package_dir = tmp.path().join("package");
    let docs_path = opts
        .docs_path
        .clone()
        .or_else(|| detect_docs_path(&package_dir, NPM_DOC_CANDIDATES))
        .ok_or_else(|| {
            anyhow::anyhow!(
                "No docs found in {spec}. Specify --docs-path to point to the docs directory \
                 within the package."
            )
        })?;

    let docs_dir = package_dir.join(&docs_path);
    // Containment guard: a `--docs-path` that is absolute, uses `../`, or points
    // at a symlink escaping the package must not let ask read arbitrary local
    // paths. Mirror try_local_read's lexical + realpath checks (the local-first
    // path had them; the tarball path did not).
    if !lexical_clean(&docs_dir).starts_with(lexical_clean(&package_dir)) {
        bail!("Docs path \"{docs_path}\" escapes the package directory in {spec}.");
    }
    if !docs_dir.exists() {
        bail!(
            "Docs path \"{docs_path}\" not found in {spec}. Available paths:\n{}",
            list_dirs(&package_dir)
        );
    }
    let real_pkg = std::fs::canonicalize(&package_dir)?;
    let real_docs = std::fs::canonicalize(&docs_dir)?;
    if !real_docs.starts_with(&real_pkg) {
        bail!("Docs path \"{docs_path}\" resolves outside the package directory in {spec}.");
    }

    let files = if docs_dir.is_file() {
        let content = std::fs::read_to_string(&docs_dir)?;
        let name = docs_dir
            .file_name()
            .map(|n| n.to_string_lossy().into_owned())
            .unwrap_or_default();
        vec![DocFile {
            path: name,
            content,
        }]
    } else {
        collect_doc_files(&docs_dir, &docs_dir)?
    };

    write_to_store(ask_home, pkg, &resolved_version, &files)?;

    Ok(FetchResult {
        files,
        resolved_version: resolved_version.clone(),
        store_path: Some(npm_store_path(ask_home, pkg, &resolved_version)?),
        store_subpath: None,
        from_store_cache: false,
        meta: FetchMeta {
            tarball: Some(tarball_url),
            integrity,
            ..Default::default()
        },
    })
}

fn list_dirs(dir: &Path) -> String {
    let Ok(entries) = std::fs::read_dir(dir) else {
        return String::new();
    };
    let mut names: Vec<String> = entries
        .flatten()
        .filter(|e| e.file_type().map(|t| t.is_dir()).unwrap_or(false))
        .map(|e| format!("  - {}/", e.file_name().to_string_lossy()))
        .collect();
    names.sort();
    names.join("\n")
}

/// Write fetched docs into `<ASK_HOME>/npm/<pkg>@<version>/` (skip if present).
fn write_to_store(ask_home: &Path, pkg: &str, version: &str, files: &[DocFile]) -> Result<()> {
    let store_dir = npm_store_path(ask_home, pkg, version)?;
    if store_dir.exists() {
        return Ok(());
    }
    if let Some(lock) = acquire_entry_lock(&store_dir)? {
        write_entry_atomic(&store_dir, &to_entry_files(files))?;
        stamp_entry(&store_dir)?;
        drop(lock);
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn version_matches_node_semver_semantics() {
        assert!(version_matches("latest", "9.9.9"));
        // Bare version → exact.
        assert!(version_matches("1.2.3", "1.2.3"));
        assert!(!version_matches("1.2.3", "1.2.4"));
        // Range → satisfies.
        assert!(version_matches("^1.2.0", "1.5.0"));
        assert!(!version_matches("^1.2.0", "2.0.0"));
        assert!(version_matches("~3.22.0", "3.22.9"));
        // Opaque tag → exact.
        assert!(version_matches("canary", "canary"));
        assert!(!version_matches("canary", "1.0.0"));
    }

    fn setup_pkg(project: &Path, pkg: &str, version: &str) -> std::path::PathBuf {
        let pkg_dir = project.join("node_modules").join(pkg);
        std::fs::create_dir_all(&pkg_dir).unwrap();
        std::fs::write(
            pkg_dir.join("package.json"),
            format!(r#"{{"name":"{pkg}","version":"{version}"}}"#),
        )
        .unwrap();
        pkg_dir
    }

    #[test]
    fn local_read_returns_docs_when_version_matches() {
        let dir = tempfile::tempdir().unwrap();
        let pkg_dir = setup_pkg(dir.path(), "acme", "2.0.0");
        std::fs::create_dir_all(pkg_dir.join("dist/docs")).unwrap();
        std::fs::write(pkg_dir.join("dist/docs/a.md"), "A").unwrap();

        let result = try_local_read(dir.path(), "acme", "2.0.0", Some("dist/docs")).unwrap();
        assert_eq!(result.resolved_version, "2.0.0");
        assert_eq!(
            result.files,
            vec![DocFile {
                path: "a.md".into(),
                content: "A".into()
            }]
        );
        assert_eq!(result.meta.install_path.as_deref(), Some(pkg_dir.as_path()));
    }

    #[test]
    fn local_read_none_without_docs_path() {
        let dir = tempfile::tempdir().unwrap();
        setup_pkg(dir.path(), "acme", "2.0.0");
        assert!(try_local_read(dir.path(), "acme", "2.0.0", None).is_none());
    }

    #[test]
    fn local_read_none_on_version_mismatch() {
        let dir = tempfile::tempdir().unwrap();
        let pkg_dir = setup_pkg(dir.path(), "acme", "1.0.0");
        std::fs::create_dir_all(pkg_dir.join("docs")).unwrap();
        std::fs::write(pkg_dir.join("docs/a.md"), "A").unwrap();
        assert!(try_local_read(dir.path(), "acme", "2.0.0", Some("docs")).is_none());
    }

    #[test]
    fn local_read_none_on_traversal_docs_path() {
        let dir = tempfile::tempdir().unwrap();
        setup_pkg(dir.path(), "acme", "2.0.0");
        assert!(try_local_read(dir.path(), "acme", "2.0.0", Some("../../etc")).is_none());
    }

    #[test]
    fn local_read_none_when_empty_docs_dir() {
        let dir = tempfile::tempdir().unwrap();
        let pkg_dir = setup_pkg(dir.path(), "acme", "2.0.0");
        std::fs::create_dir_all(pkg_dir.join("docs")).unwrap();
        // No doc files inside → treated as a miss.
        assert!(try_local_read(dir.path(), "acme", "2.0.0", Some("docs")).is_none());
    }

    #[test]
    fn local_read_single_file_docs_path() {
        let dir = tempfile::tempdir().unwrap();
        let pkg_dir = setup_pkg(dir.path(), "acme", "2.0.0");
        std::fs::write(pkg_dir.join("README.md"), "readme").unwrap();
        let result = try_local_read(dir.path(), "acme", "2.0.0", Some("README.md")).unwrap();
        assert_eq!(
            result.files,
            vec![DocFile {
                path: "README.md".into(),
                content: "readme".into()
            }]
        );
    }

    #[test]
    fn fetch_writes_local_read_into_store() {
        let dir = tempfile::tempdir().unwrap();
        let ask_home = dir.path().join("ask-home");
        let pkg_dir = setup_pkg(dir.path(), "acme", "2.0.0");
        std::fs::create_dir_all(pkg_dir.join("docs")).unwrap();
        std::fs::write(pkg_dir.join("docs/a.md"), "A").unwrap();

        let client = crate::http::mock::MockClient::new();
        let opts = NpmOptions {
            name: "acme".into(),
            version: "2.0.0".into(),
            docs_path: Some("docs".into()),
            ..Default::default()
        };
        let result = fetch(&client, &opts, dir.path(), &ask_home).unwrap();
        let store = result.store_path.unwrap();
        assert!(store.ends_with("npm/acme@2.0.0"));
        assert!(store.join("a.md").exists());
    }
}
